distribution/lib/Standard/Table/0.0.0-dev/src/Internal/Delimited_Reader.enso

```diff
@@ -101,49 +106,92 @@ read_from_reader format java_reader on_problems max_columns=4096 =
         Nothing -> -1
         Integer -> format.row_limit
         _ -> Error.throw (Illegal_Argument_Error "`row_limit` should be Integer or Nothing.")
+    warnings_as_errors = on_problems == Problem_Behavior_Module.Report_Error
+    quote_characters = case format.quote_style of
+        Quote_Style.No_Quotes -> Pair Nothing Nothing
+        Quote_Style.With_Quotes _ quote quote_escape -> Pair quote quote_escape
+    base_parser = case format.quote_style of
+        Quote_Style.No_Quotes -> IdentityParser.new
+        Quote_Style.With_Quotes _ quote _ ->
+            QuoteStrippingParser.new quote
+    value_parser = if format.value_formatter.is_nothing then base_parser else
+        wrapped = format.value_formatter.wrap_base_parser base_parser
+        TypeInferringParser.new format.value_formatter.get_specific_type_parsers.to_array wrapped
+    cell_type_guesser = if format.headers != Infer then Nothing else
+        formatter = format.value_formatter.if_nothing Data_Formatter
+        TypeInferringParser.new formatter.get_specific_type_parsers.to_array IdentityParser.new
+    DelimitedReader.new java_reader format.delimiter quote_characters.first quote_characters.second java_headers format.skip_rows row_limit max_columns value_parser cell_type_guesser format.keep_invalid_rows warnings_as_errors
+
+translate_reader_problem problem =
+    invalid_row = [InvalidRow, (java_problem-> Invalid_Row java_problem.source_row java_problem.table_index (Vector.Vector java_problem.row))]
+    additional_invalid_rows = [AdditionalInvalidRows, (java_problem-> Additional_Invalid_Rows java_problem.count)]
+    mismatched_quote = [MismatchedQuote, (_-> Mismatched_Quote)]
+    duplicate_names = [DuplicateNames, (java_problem-> Duplicate_Output_Column_Names (Vector.Vector java_problem.duplicatedNames))]
+    invalid_names = [InvalidNames, (java_problem-> Invalid_Output_Column_Names (Vector.Vector java_problem.invalidNames))]
+
+    translations = [invalid_row, additional_invalid_rows, mismatched_quote, duplicate_names, invalid_names]
+    found = translations.find t-> Java.is_instance problem t.first
+    if found.is_error then problem else
+        found.second problem
+
+## PRIVATE
+   An internal type representing columns deduced from an existing file.
+type Detected_Headers
+    ## Indicates that the file did not exist or was empty.
+    type No_Data
+
+    ## Represents the headers found in the file.
+    type Existing_Headers (column_names : Vector Text)
+
+    ## Indicates that the file exists but no headers have been found, so only positional column matching is possible.
+    type No_Headers (column_count : Integer)
 
+## PRIVATE
+   Reads the beginning of the file to detect the existing headers and column
+   count.
+detect_headers : File -> File_Format.Delimited -> Detected_Headers
```

Comment:
It kinda feels like this should use a `Maybe` and be empty instead of using a separate new marker type `No_Data`

Reply:
Hmm, I thought it makes sense to create a domain-specific ADT for all the cases here. But your solution also sounds good to me, so I can switch to it. I think I'd use a `Nothing` as we know it would either be a `Nothing` or one of the two other ADT cases - using `Maybe` seems like an overkill to me because I want to represent 3 states and not a presence or lack of two states.